File: Problems 1800-1899/1857. Largest Color Value in a Directed Graph/solution.cpp
class Solution {
    public:
        int largestPathValue(string colors, vector<vector<int>>& edges) {
            int n = colors.length();
            vector<vector<int>> graph(n);
            vector<int> indegrees(n, 0);
            for (auto& edge : edges) {
                int a = edge[0];
                int b = edge[1];
                graph[a].push_back(b);
                indegrees[b]++;
            }
    
            // dp[i][j]: longest number of nodes with color j on path ending at i
            vector<vector<int>> dp(n, vector<int>(26, 0));
            queue<int> q;
            for (int node = 0; node < n; node++) {
                if (indegrees[node] == 0) {
                    q.push(node);
                    // since node is starting point (indegree of 0),
                    // its longest length path ending at itself it 1
                    dp[node][colors[node]-'a'] = 1;
                }
            }
    
            int res = 0;
            int seen = 0;
            while (!q.empty()) {
                int node = q.front();
                q.pop();
                seen++;
                res = max(res, *max_element(dp[node].begin(), dp[node].end()));
    
                for (int adj : graph[node]) {
                    for (int color = 0; color < 26; color++) {
                        // dp[adj][color] = max(dp[adj][color], dp[node][color] + (colors[adj]-'a' == color));
                        dp[adj][color] = max(dp[adj][color], dp[node][color]);
                    }
                    indegrees[adj]--;
                    if (indegrees[adj] == 0) {
                        q.push(adj);
                        dp[adj][colors[adj]-'a']++;
                    }
                }
            }
    
            return seen == n ? res : -1;
        }
    };

File: Problems 1800-1899/1857. Largest Color Value in a Directed Graph/solution.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        graph = [[] for _ in range(n)]
        for a, b in edges:
            graph[a].append(b)

        # dp[i][j] -> number of j color in path starting at i
        dp = [[0]*26 for _ in range(n)]
        visited = [0]*n

        res = 0
        for node in range(n):
            if self.solve(colors, node, graph, dp, visited):
                return -1
            res = max(res, max(dp[node]))
        return res
        
    def solve(self, colors, node, graph, dp, visited):
        if visited[node] == -1:
            return True
        
        if visited[node] == 1:
            return False
        
        visited[node] = -1
        for adj in graph[node]:
            if self.solve(colors, adj, graph, dp, visited):
                return True
            for color in range(26):
                dp[node][color] = max(dp[node][color], dp[adj][color])

        visited[node] = 1
        dp[node][ord(colors[node])-ord('a')] += 1
        return False